src/python/py/models/builder.py

```diff
@@ -1495,6 +1506,125 @@ def make_mlp_fc(self, layer_id, mlp, root_input):
+            import tensorrt_llm
+
+            _, processed_q_weight, torch_weight_scales = (
+                torch.ops.trtllm._symmetric_quantize_last_axis_of_batched_matrix(weights.T.cpu().contiguous(), type)
+            )
+
+            return torch_weight_scales.to(torch.float16), processed_q_weight
+
+        w1_list = []
+        w2_list = []
+        w3_list = []
+        w1_scale_list = []
+        w2_scale_list = []
+        w3_scale_list = []
+
+        for i in range(num_experts):
+            # Quantize the weights with uint8
+            w1_scale, pre_qweight1= quant_dequant(bsm.experts[i].w1.weight, use_int4)
+            w2_scale, pre_qweight2= quant_dequant(bsm.experts[i].w2.weight, use_int4)
+            w3_scale, pre_qweight3= quant_dequant(bsm.experts[i].w3.weight, use_int4)
+
+            w1_list.append(pre_qweight1)
+            w2_list.append(pre_qweight2)
+            w3_list.append(pre_qweight3)
+
+            w1_scale_list.append(w1_scale)
+            w2_scale_list.append(w2_scale)
+            w3_scale_list.append(w3_scale)
+
+        moe_expert_weight_1_name = f"model.layers.{layer_id}.moe.weight_1"
+        moe_expert_weight_2_name = f"model.layers.{layer_id}.moe.weight_2"
+        moe_expert_weight_3_name = f"model.layers.{layer_id}.moe.weight_3"
+
+        moe_expert_scales_1_name = f"model.layers.{layer_id}.moe.scales_1"
+        moe_expert_scales_2_name = f"model.layers.{layer_id}.moe.scales_2"
+        moe_expert_scales_3_name = f"model.layers.{layer_id}.moe.scales_3"
+
+        def make_moe_external_tensor(w_list, moe_expert_name, numpy_type):
+            moe_experts_weight = torch.stack(w_list, dim=0).detach().numpy()
+            self.make_external_tensor(moe_experts_weight.astype(numpy_type), moe_expert_name)
+
+        make_moe_external_tensor(w1_list, moe_expert_weight_1_name, np.uint8)
+        make_moe_external_tensor(w2_list, moe_expert_weight_2_name, np.uint8)
+        make_moe_external_tensor(w3_list, moe_expert_weight_3_name, np.uint8)
+
+        # Currently we don't expect QMoE to be used with distributed inference
+        make_moe_external_tensor(w1_scale_list, moe_expert_scales_1_name, self.to_numpy_dtype[self.io_dtype])
+        make_moe_external_tensor(w2_scale_list, moe_expert_scales_2_name, self.to_numpy_dtype[self.io_dtype])
+        make_moe_external_tensor(w3_scale_list, moe_expert_scales_3_name, self.to_numpy_dtype[self.io_dtype])
+
+        bias_ph = "" # Placeholder for bias
+        inputs = [root_input, f"{gate_reshape_name}/output_0", \
+                  moe_expert_weight_1_name, moe_expert_scales_1_name, bias_ph, \
+                  moe_expert_weight_2_name, moe_expert_scales_2_name, bias_ph, \
+                  moe_expert_weight_3_name, moe_expert_scales_3_name]
+        output = f"{moe_name}/output_0"
+
+        op_type = "QMoE"
+        self.make_node(op_type, inputs=inputs, outputs=[output], name=moe_name, domain="com.microsoft",
```

Comment:
Can we create two generic methods called `make_moe` and `make_qmoe` and call those instead? We can use the logic similar to how `make_mlp` works.

Reply:
will refactor it when adding the sharded moe part